test/model/message_list_test.dart

```diff
@@ -0,0 +1,241 @@
+      const newTimestamp = 99999;
+
+      List<String> oldFlags = [];
+      List<String> newFlags = ["starred"];
+
+      final mockMessage = eg.streamMessage(id: 243, stream: stream, content: oldContent, flags: oldFlags);
+      final messageList = await messageListViewWithMessages([mockMessage], store, narrow);
+
+      final updateEvent = UpdateMessageEvent(
+        id: 1,
+        messageId: mockMessage.id,
+        messageIds: [mockMessage.id],
+        flags: newFlags,
+        renderedContent: newContent,
+        editTimestamp: newTimestamp,
+        isMeMessage: true,
+        userId: userId
+      );
+
+      final message = messageList.messages[0];
+      check(message)
+        ..content.equals(oldContent)
+        ..flags.deepEquals(oldFlags)
+        ..isMeMessage.equals(false);
+
+      var listenersNotified = false;
+
+      messageList.addListener(() { listenersNotified = true; });
+      messageList.maybeUpdateMessage(updateEvent);
+
+      final updatedMessage = messageList.messages[0];
+      check(updatedMessage).identicalTo(message);
+      check(listenersNotified).equals(true);
+
+      check(message)
+        ..content.equals(newContent)
+        ..lastEditTimestamp.equals(newTimestamp)
+        ..flags.equals(newFlags)
+        ..isMeMessage.equals(true);
+    });
+
+    test('update event is ignored when message is not in the message list', () async {
+      final store = await setupStore(stream);
+
+      const oldContent = "<p>Hello, world</p>";
+      const newContent = "<p>Hello, edited</p>";
+      const newTimestamp = 99999;
+
+      final mockMessage = eg.streamMessage(id: 243, stream: stream, content: oldContent);
+      final messageList = await messageListViewWithMessages([mockMessage], store, narrow);
+
+      final updateEvent = UpdateMessageEvent(
+        id: 1,
+        messageId: 972,
+        messageIds: [972],
+        flags: mockMessage.flags,
+        renderedContent: newContent,
+        editTimestamp: newTimestamp,
+        userId: userId,
+      );
+
+      final message = messageList.messages[0];
+      check(message).content.equals(oldContent);
+
+      var listenersNotified = false;
+
+      messageList.addListener(() { listenersNotified = true; });
+      messageList.maybeUpdateMessage(updateEvent);
+
+      check(listenersNotified).equals(false);
+      check(message).content.equals(oldContent);
+
+    });
+    test('rendering-only update does not change timestamp', () async {
+      final store = await setupStore(stream);
+
+      const oldContent = "<p>Hello, world</p>";
+      const newContent = "<p>Hello, world</p> <div>Some link preview</div>";
+      const newTimestamp = 99999;
+
+      final mockMessage = eg.streamMessage(id: 972, stream: stream, content: oldContent);
+      final messageList = await messageListViewWithMessages([mockMessage], store, narrow);
+
+      final updateEvent = UpdateMessageEvent(
+        id: 1,
+        messageId: 972,
+        messageIds: [972],
```

Comment:
I like the approach you took in one of the test cases above: instead of writing out a magic number as message ID that crucially happens to be the same number as used in the original message, refer directly to the original message object and its `id` property. That helps make it clear that the point is it's the same message.